[DOCS] DOC-362: how to choose between working with a single or multiple batches of data

Rachel-Reverie opened the pull request:
DOC-362
Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

Changes proposed in this pull request:
- New documentation on how to choose between working with a single or multiple batches of data written and added to sidebar


After submitting your PR, CI checks will run and @cla-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-


### Definition of Done
Please delete options that are not relevant.

- [X] I have made corresponding changes to the documentation

Thank you for submitting!

Rachel-Reverie committed:
DOC-362
- First draft

Rachel-Reverie committed:
DOC-362
- Minor edits to polish content
- refactor of file name to match doc title

ghost commented:
<details>
<summary><b>👇 <a href="https://app.codesee.io/r/reviews?pr=5745&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations">Click on the image</a> for a new way to code review</b>
</summary>

  - Make big changes easier — review code in small groups of related files
  - Know where to start — see the whole change at a glance
  - Take a code tour — explore the change with an interactive tour
  - Make comments and review — all fully sync’ed with github

    [Try it now!](https://app.codesee.io/r/reviews?pr=5745&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations)
</details>

[![Review these changes using an interactive CodeSee Map](https://s3.us-east-2.amazonaws.com/maps.codesee.io/images/github/great-expectations/great_expectations/5745/adcd01ca/e5044ac9b471ce16239a052fa715c77cc2abfb78.svg)](https://app.codesee.io/r/reviews?pr=5745&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations)

#### Legend
![CodeSee Map Legend](https://app.codesee.io/Legend_v2.png)

Shinnnyshinshin commented on docs/guides/connecting_to_your_data/how_to_choose_between_working_with_a_single_or_multiple_batches_of_data.md:
```diff
@@ -0,0 +1,108 @@
+In some cases, a single Batch of data is all you need (or all that is supported).  In particular:
+- When you are Validating data, you will need to specify a single Batch of data.  If you provide multiple Batches in a Batch Request, Validations will default to operating on the last Batch in the list.
+- The Runtime Data Connector does not support handling multiple Batches of data.  You will always configure your Data Assets to correspond to at most a single Batch of Data when you are using a Runtime Data Connector.
+- If you are creating Expectations that don't rely on per-Batch statistical analysis, or you want to quickly create some Expectations as part of an early Expectation Suite or exploratory data analysis, a single Batch is likely to be all that you need.  
```
I think this is a really good point, and may be worth discussing further.  Just rattling off a few points: 
- One of the selling points of the `DataAssistant` and `RBP` is the fact that it **can** do multi-batch statistical analysis (in addition to simple "range" calculations)... and to my knowledge, this wasn't really possible before the era of RBPs
- What do you think about switching the order of this statement? Something like 

If you want to quickly create some Expectations as part of an early Expectation Suite or exploratory data analysis, a single Batch is likely to be all that you need.  If you are creating Expectations that require parameters that should be calculated across multiple Batches, you will need a multi-batch configuration. 

Or even dropping the second sentence ("If you are creating Expectations...) entirely since we are going into more detail below?  (I feel like this might be a better option after looking at the section below)

Rachel-Reverie replied:
I updated the last bullet point according to your suggestions.

Shinnnyshinshin commented on docs/guides/connecting_to_your_data/how_to_choose_between_working_with_a_single_or_multiple_batches_of_data.md:
```diff
@@ -0,0 +1,108 @@
+
+:::note 
+
+When Validating data you may include multiple *Batch Requests* in a single Checkpoint.  But only one Batch out of each Batch Request will be used in the Validation process.
```
Love it.  Having gone through this process of writing the doc, how hard was this to wrap your mind around? Would a visual help here?

Rachel-Reverie replied:
I would like to add more visuals to the docs just in general and this is a good candidate for one.  I'm going to create a ticket for adding one here -- but that will be independent of this PR, so that the rest of the Data Assistant doc work doesn't get bogged down.

Rachel-Reverie replied:
Wrapping my mind around this facet was pretty easy, though.  Mostly because I did a _lot_ of clarifying on the `validations` parameter of a BatchRequest previously, when working on documentation about passing additional Expectation Suites and Batch Requests into a Checkpoint.  So I already had a firm grounding in the principle of multiple BatchRequests going into a Checkpoint.

Shinnnyshinshin commented on docs/guides/connecting_to_your_data/how_to_choose_between_working_with_a_single_or_multiple_batches_of_data.md:
```diff
@@ -0,0 +1,108 @@
+
+This is particularly true if you are working with Expectations that involve statistical analysis on a per-Batch basis.
+
+For example, let's say you are working with the NYC taxi data and intended to define an Expectation to validate that the median trip distance of rides taken within a given month falls in a specific range. You would first need to ensure you could divide your data up into months for analysis, which you could do by configuring your Datasource to define an individual Data Asset for each month.  Then, ...
```
I feel like this is a critical point too. Do you think it would be strengthened with a more specific description? 

* What range does the median `fare` fall between for June and December? Next, can we write an Expectation that alters us if the value falls out of that range?

Shinnnyshinshin commented on docs/guides/connecting_to_your_data/how_to_choose_between_working_with_a_single_or_multiple_batches_of_data.md:
```diff
@@ -0,0 +1,108 @@
+
+Instead, you can configure your Datasource to define a single Data Asset that groups each month's worth of data into a discrete Batch.  Then, you can use a single Batch Request to retrieve that list of Batches. By feeding that Batch Request into the auto-initializing Expectation `expect_column_median_to_be_between` and setting the parameters `(column="trip_distance", auto=True)` you can have Grea ...
+
+Alternatively, if you know you will need additional Expectations, you can feed that Batch Request into a Data Assistant, which will generate an entire Expectation Suite with Expectations that have had their parameters determined from the analysis of the provided Batches... saving you from even more repetitive work. Huzzah for automation!
```
❤️

Shinnnyshinshin reviewed (commented):
Thank you @Rachel-Reverie. Left some questions in-line, and a larger question around the Filesystem/ Spark / Sql examples.

* Is there a place for them here? 
* Or were you thinking of another doc? 
* Or are we leaving them as reference? 

Thanks again :)

Rachel-Reverie commented:
> Filesystem/ Spark / Sql examples.

The current plan is for those examples to be included in additional docs.  Since the formatting of each is a bit different, I plan to write an individual guide for each of: Pandas, PySpark, and Sql Datasources.  Each of those individual guides will have an example of how to configure Data Assets that only return a single Batch, or may return multiple Batches (if supported by the Data Connector), for each of: the inferred, configured, and runtime Data Connectors.  Once those guides are up, I'll add them all as links under the "Next steps" header on this doc.

Rachel-Reverie committed:
DOC-362
- Update last bullet point of "when to work with a single Batch of data" to be more concise.

Shinnnyshinshin commented on docs/guides/connecting_to_your_data/how_to_choose_between_working_with_a_single_or_multiple_batches_of_data.md:
```diff
@@ -0,0 +1,108 @@
+
+In some cases, a single Batch of data is all you need (or all that is supported).  In particular:
+- When you are Validating data, you will need to specify a single Batch of data.  If you provide multiple Batches in a Batch Request, Validations will default to operating on the last Batch in the list.
+- The Runtime Data Connector does not support handling multiple Batches of data.  You will always configure your Data Assets to correspond to at most a single Batch of Data when you are using a Runtime Data Connector.
+- If you want to quickly create some Expectations as part of an early Expectation Suite or exploratory data analysis, a single Batch is likely to be all that you need.
```
Something for my edification, but when do we use `RuntimeDataConnector` vs Runtime Data Connector? Would it only be `RuntimeDataConnector` when we are describing code directly?

Rachel-Reverie replied:
Yes, if we're referring to a code snippet we would want to use the actual class name.  But if we're just talking about it conceptually, we would just use the name of the concept.

Shinnnyshinshin commented on docs/guides/connecting_to_your_data/how_to_choose_between_working_with_a_single_or_multiple_batches_of_data.md:
```diff
@@ -0,0 +1,108 @@
+
+For example, let's say you are working with the NYC taxi data and intended to define an Expectation to validate that the median trip distance of rides taken within a given month falls in a specific range. You would first need to ensure you could divide your data up into months for analysis, which you could do by configuring your Datasource to define an individual Data Asset for each month.  Then, ...
+
+Instead, you can configure your Datasource to define a single Data Asset that groups each month's worth of data into a discrete Batch.  Then, you can use a single Batch Request to retrieve that list of Batches. By feeding that Batch Request into the auto-initializing Expectation `expect_column_median_to_be_between` and setting the parameters `(column="trip_distance", auto=True)` you can have Grea ...
```
Love this

Shinnnyshinshin commented on docs/guides/connecting_to_your_data/how_to_choose_between_working_with_a_single_or_multiple_batches_of_data.md:
```diff
@@ -0,0 +1,108 @@
+
+The reason that this case only applies to Sql Datasources is due to the difference in how Batches are handled in File System Datasources and Sql Datasources.
+
+#### How do Sql Datasources and File System Datasources differ in how they handle Batches?
+
```
Love this section

Shinnnyshinshin reviewed (approved):
Content looks wonderful @Rachel-Reverie,  I'm approving with a non-blocking question: 
* "Do we intend" or "where do we intend" to include code snippets of configurations that will return single or multi-batch?

Rachel-Reverie commented:
> Content looks wonderful @Rachel-Reverie, I'm approving with a non-blocking question:
> 
> * "Do we intend" or "where do we intend" to include code snippets of configurations that will return single or multi-batch?

Each of the Filesystem/ Spark / Sql Datasource docs will include examples of both the configuration for a Data Asset that can only return a single Batch, and the configuration for a Data Asset that is capable of returning multiple Batches.  This document is just meant to explain when each configuration is best, but those docs will be focused on the process of actually implementing desired configurations.

Rachel-Reverie committed:
Merge branch 'develop' into DOCS/DOC-362/how_to_choose_between_single_and_multi_batch_datasource_configurations

github-actions[bot] committed:
Merge develop into DOCS/DOC-362/how_to_choose_between_single_and_multi_batch_datasource_configurations

github-actions[bot] committed:
Merge develop into DOCS/DOC-362/how_to_choose_between_single_and_multi_batch_datasource_configurations

github-actions[bot] committed:
Merge develop into DOCS/DOC-362/how_to_choose_between_single_and_multi_batch_datasource_configurations

github-actions[bot] committed:
Merge develop into DOCS/DOC-362/how_to_choose_between_single_and_multi_batch_datasource_configurations

donaldheppner reviewed (approved):
🎉
